src/core/unit.c

```diff
@@ -653,6 +645,12 @@ void unit_free(Unit *u) {
         for (d = 0; d < _UNIT_DEPENDENCY_MAX; d++)
                 bidi_set_free(u, u->dependencies[d]);
 
+        /* A unit is being dropped from the tree, make sure our siblings and ancestor slices are realized
+         * properly. Do this after we detach the unit from slice tree in order to eliminate its effect on
+         * controller masks. */
+        if (u->type != _UNIT_TYPE_INVALID && UNIT_HAS_CGROUP_CONTEXT(u))
+                unit_add_siblings_to_cgroup_realize_queue(u);
```

Comment:
the comment is misleading: the slice is still semi-attached to the unit, the `unit_ref_unset(&u->slice);` call comes a few lines further down only.

I'd prefer if we'd change this around a bit, so that we don't call unit_add_siblings_to_cgroup_realize_queue() on a semi-destroyed unit (which `u` is at this point.)

maybe we should split unit_add_siblings_to_cgroup_realize_queue() up, and have

```c
void slice_add_family_to_cgroup_realize_queue(Slice *slice) {
        /* We define "family" as the slice's immediate children, the slice itself, and all parents all the way up the tree, as well as all siblings of these. */
        …
}
void unit_add_siblings_and_parents_to_cgroup_realize_queue(u) {
        slice_add_family_to_cgroup_realize_queue(u->slice);
}
```

and then call slice_add_family_to_cgroup_realize_queue(…) here on the slice object, after we *fully* detached our unit from it. That way things should be more robust, as we won't run into trouble with "half-destroyed" objects ever.

Reply:
if you do it like that then you can drop the special "self" handling you added to  unit_add_siblings_and_parents_to_cgroup_realize_queue(), right?

Reply:
Naming the set we need to operate on is a good idea. I introduced family only (no siblings function) and explained in the comments its effect and purpose.

I've left the call-site in `unit_free` the same. There is (and was) a "race" between slice GC queue processing and cgroup realization queue. That should be OK:
- if GC goes before CG, then the unit is removed from CG queue (but it should have been `unit_prune_cgroup`'d by then) and ancestors are still CG queued,
- if CG goes before GC, then it's like the normal case (i.e. when no GC happened).